### PR TITLE
WMDP-126: Add implementation of the mock endpoint

### PR DESCRIPTION
**Pull request by chouinar**
## Ticket

https://wicmtdp.atlassian.net/browse/wmdp-126

## Changes
Added the bulk of the implementation (besides clinic selection) to the mock API

Added several overrides/adjustment validators/formatters for the request/response objects to make them consistent & more detailed.

## Context for reviewers

Rough questions / things to look out for as a reviewer:
* The allowed values for the eligibility categories + programs aren't correct. Figure we can come back once we know those values for certain and adjust them.
* I've kept validation fairly straightforward (mostly typing and nullability), if we want anything more elaborate, can add as a follow-up, but let me know if you have any ideas.
* The connexion overrides are fairly necessary to make things easy to test / easier for a user by making the error messages clearer and consistent.

## Testing

`make clean-volumes init start` to get your env running and cleaned up from any past approach.

Requests to the API endpoint:
Screen Shot 2022-09-06 at 11 30 34 AM

Response from the API endpoint:
Screen Shot 2022-09-06 at 11 30 40 AM

DB record created:
Screen Shot 2022-09-06 at 11 31 31 AM

### CODE REVIEW

File: app/api/db/migrations/alembic.ini [hunk 1]
```diff
@@ -47,7 +47,7 @@ hooks=isort, black
 
 isort.type=console_scripts
 isort.entrypoint=isort
-isort.options= --atomic --apply
+isort.options= --atomic
```
Comment (chouinar) [inline app/api/db/migrations/alembic.ini:50]: The newer version of isort doesn't require `-apply` and gives a warning if used. Had fixed in our makefile already, but saw the error during the migration.

File: app/api/util/connexion_validators.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    RequestBodyValidator,
+    ResponseBodyValidator,
+)
+from connexion.json_schema import Draft4RequestValidator, Draft4ResponseValidator
+from connexion.utils import is_null
+
+from api.util.response import ValidationErrorDetail, ValidationException
+
+
+def get_custom_validator_map():
```
Comment (rocketnova) [inline app/api/util/connexion_validators.py]: Nit: This file is missing some typing, such as the return value of this function.

File: app/api/util/connexion_validators.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    def validate_response(self, data, status_code, headers, url):
+        # Only validate json responses
+        if headers.get("Content-Type") != "application/json":
+            return True
+
+        response_body = self.operation.json_loads(data)
+
+        if url.endswith("/status"):
+            return True
```
Comment (rocketnova) [inline app/api/util/connexion_validators.py]: What does `/status` refer to? Should this be referring to the healthcheck endpoint?

File: app/api/util/response.py [hunk 4]
```diff
@@ -53,3 +75,20 @@ def success_response(
     status_code: int = 200,
 ) -> Response:
     return Response(status_code=status_code, message=message, data=data, warnings=warnings)
+
+
+def error_response(
+    status_code: HTTPException
+    | Type[HTTPException]
```
Comment (rocketnova) [inline app/api/util/response.py]: What does this typing syntax indicate? That `status_code` can either be a `HTTPException` or `Type[HTTPException]`? What's the difference between the two?

File: app/api/util/response.py [hunk 5]
```diff
@@ -53,3 +75,20 @@ def success_response(
...
+    | Type[BadRequest]
+    | Type[Conflict]
+    | Type[ServiceUnavailable]
+    | Type[NotFound]
+    | Type[Forbidden],
+    message: str,
+    errors: list[ValidationErrorDetail],
+    data: Optional[dict | list[dict]] = None,
```
Comment (rocketnova) [inline app/api/util/response.py:77]: Is this new? The ability to use type `dict` and type `list` instead of having to import them as `List` and `Dict`?

File: app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        "eligibility_screener", sa.Column("has_prior_wic_enrollment", sa.Boolean(), nullable=False)
+    )
+    op.add_column(
+        "eligibility_screener",
+        sa.Column("eligibility_programs", postgresql.ARRAY(sa.Text()), nullable=True),
+    )
+    op.add_column("eligibility_screener", sa.Column("household_size", sa.Integer(), nullable=True))
+    op.add_column("eligibility_screener", sa.Column("zip_code", sa.Text(), nullable=False))
+    op.add_column("eligibility_screener", sa.Column("applicant_notes", sa.Text(), nullable=True))
+    op.alter_column("eligibility_screener", "first_name", existing_type=sa.TEXT(), nullable=False)
```
Comment (rocketnova) [inline app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py:35]: Is there a difference between `sa.Text()` and `sa.TEXT()`?

File: app/tests/api/db/models/test_factories.py [hunk 7]
```diff
@@ -7,26 +7,51 @@
 # as the API we have implemented at the moment
 # doesn't use it.
```
Comment (rocketnova) [inline app/tests/api/db/models/test_factories.py]: Do we still need this comment?

File: app/tests/api/util/test_validators.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        "/user", json=MISSING_DATA_USER
+    ).get_json()
+    assert request_validation_error_response_no_warnings.get("warnings", None) is None
+
+    assert success_response["message"] == "Success"
+    assert success_response["data"] is not None
+    assert success_response.get("data") is not None
+
+
+def validate_invalid_response(response, message, field_prefix=""):
```
Comment (rocketnova) [inline app/tests/api/util/test_validators.py]: I find the function list in this file a little hard to follow. Could we move this helper function up to be underneath `post_user_invalid_response()`?

File: app/tests/api/util/test_validators.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+
+def get_user():
+    """handler for test api (see 'test.yml' file in this directory)"""
+    return success_response(message="Success", data=MISSING_DATA_USER).to_api_response()
+
+
+def post_user_invalid_response():
+    """handler for test api (see 'test.yml' file in this directory)"""
+    return success_response(message="Success", data=INVALID_USER).to_api_response()
```
Comment (rocketnova) [inline app/tests/api/util/test_validators.py]: Why are these success responses? Should these not be `error_response()`?

**Review by rocketnova — COMMENTED** [review]
Wow! So much great stuff. Mostly, few questions to make sure I follow and some minor suggestions. 

I have one question about validation though. If I try to post with:

```
{
  "has_prior_wic_enrollment": false,
  "household_size": "stuff",
  "last_name": "Doe",
  "phone_number": "224-705-2345",
  "zip_code": "12345"
}
```

The errors that are returned are about type, but does not include an error about missing the required field. Is this expected?

File: app/api/util/connexion_validators.py (hunk 2 above)
Comment (chouinar) [reply app/api/util/connexion_validators.py]: Hmm, why didn't mypy complain about that. Will look into + fix.

File: app/api/util/connexion_validators.py (hunk 3 above)
Comment (chouinar) [reply app/api/util/connexion_validators.py]: Yes, missed changing this in the copy over. Good catch.

File: app/api/util/response.py (hunk 4 above)
Comment (chouinar) [reply app/api/util/response.py]: So the various `Type[...]` are because you normally call this like `error_response(BadRequest, ..)` and are just passing in the class which each of these has the status code as a static value. But sometimes you might pass in an error that was actually thrown so you have an object, not the class itself. Both have the `status_code` value, so it's fine to use either.

I thought MyPy couldn't handle `Type[HTTPException]` for derived classes (maybe an older version couldn't?), but it looks like it's fine with that, so can simplify it.

File: app/api/util/response.py (hunk 5 above)
Comment (chouinar) [reply app/api/util/response.py:77]: Yep, Python 3.9 I believe added it which saves having to import the annotation from the typing library. 3.10 added the `A | B` syntax in place of `Union[A, B]`.

File: app/api/util/connexion_validators.py (hunk 2 above)
Comment (chouinar) [reply app/api/util/connexion_validators.py]: Apparently an entirely untyped function doesn't get checked by default: https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-every-function-has-an-annotation-no-untyped-def

File: app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py (hunk 6 above)
Comment (chouinar) [reply app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py:35]: Looks like `TEXT` is a subclass of `Text` with a different `__visit_name__` and no other difference. Not sure if SQLAlchemy treats those differently in some way? I see the same pattern in the Mass migrations where alter column always did `TEXT` and add column always did `Text`. So, should be fine?

File: app/tests/api/util/test_validators.py (hunk 8 above)
Comment (chouinar) [reply app/tests/api/util/test_validators.py]: Yes, I wonder if we can make a linter utility that makes sure all functions not starting with `test_` come before those that do start with `test_`.

File: app/tests/api/util/test_validators.py (hunk 9 above)
Comment (chouinar) [reply app/tests/api/util/test_validators.py:36]: Hmm, that is odd. Will fix.

Comment (chouinar) [reply app/tests/api/util/test_validators.py]: That is odd, fixed it to be an error.

Comment (chouinar) [reply app/tests/api/util/test_validators.py]: Looks like there is a bit of extra/unused pieces here, will clean up.

**Comment by chouinar** [comment]
> 



> Wow! So much great stuff. Mostly, few questions to make sure I follow and some minor suggestions.
> 
> I have one question about validation though. If I try to post with:
> 
> ```
> {
>   "has_prior_wic_enrollment": false,
>   "household_size": "stuff",
>   "last_name": "Doe",
>   "phone_number": "224-705-2345",
>   "zip_code": "12345"
> }
> ```
> 
> The errors that are returned are about type, but does not include an error about missing the required field. Is this expected?

Hmm, this is due to how it's currently handled. We have validation in two places, the specification in openapi and the pydantic. Openapi happens first and if it encounters any errors, just returns those (going through some of our reformatting/logging). If it passes, then it hits the pydantic errors.

So, the type checking is happening in your request at the openapi layer, but we don't have anything marked as required at the moment on the request and pydantic handles that. This is the pattern Mass had, but we could add that to openapi so they all happen at once.

File: app/api/util/connexion_validators.py (hunk 2 above)
Comment (rocketnova) [reply app/api/util/connexion_validators.py]: 👀 interesting! Can we turn that on?

File: app/api/util/response.py (hunk 4 above)
Comment (rocketnova) [reply app/api/util/response.py]: Ah I see. Thank you for explaining and simplifying.

File: app/api/util/response.py (hunk 5 above)
Comment (rocketnova) [reply app/api/util/response.py:77]: Yay!

File: app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py (hunk 6 above)
Comment (rocketnova) [reply app/api/db/migrations/versions/2022_09_02_12_43_41_7b9a08694907_wmdp_126_add_screener_model.py:35]: 🤔 Weird. 🤷 Happy to resolve this comment.

File: app/tests/api/util/test_validators.py (hunk 8 above)
Comment (rocketnova) [reply app/tests/api/util/test_validators.py]: Oooo that would be amazing

**Comment by rocketnova** [comment]
> > 
> 
> > Wow! So much great stuff. Mostly, few questions to make sure I follow and some minor suggestions.
> > I have one question about validation though. If I try to post with:
> > ```
> > {
> >   "has_prior_wic_enrollment": false,
> >   "household_size": "stuff",
> >   "last_name": "Doe",
> >   "phone_number": "224-705-2345",
> >   "zip_code": "12345"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The errors that are returned are about type, but does not include an error about missing the required field. Is this expected?
> 
> Hmm, this is due to how it's currently handled. We have validation in two places, the specification in openapi and the pydantic. Openapi happens first and if it encounters any errors, just returns those (going through some of our reformatting/logging). If it passes, then it hits the pydantic errors.
> 
> So, the type checking is happening in your request at the openapi layer, but we don't have anything marked as required at the moment on the request and pydantic handles that. This is the pattern Mass had, but we could add that to openapi so they all happen at once.

Ah I suspected that it was something like thate where we are doing two layes of checking and returning early if the first layer fails. This probably doesn't surprise you since I think I made a similar request with payments stuff, but I think it's a better API quality of life if all the errors are returned, rather than a subset.

File: app/tests/api/util/test_validators.py (hunk 8 above)
Comment (chouinar) [reply app/tests/api/util/test_validators.py]: I'll start making a list of ideas to put in the generic platform if possible.

File: app/api/util/connexion_validators.py (hunk 2 above)
Comment (chouinar) [reply app/api/util/connexion_validators.py]: Yes, can add `disallow_untyped_defs = true` to the mypy config in the pyproject file.

But as it causes 34 errors, lets make that a follow-up (will add to my follow-up/cleanup tasks).

**Comment by chouinar** [comment]
> > > 
> > 
> > 
> > > Wow! So much great stuff. Mostly, few questions to make sure I follow and some minor suggestions.
> > > I have one question about validation though. If I try to post with:
> > > ```
> > > {
> > >   "has_prior_wic_enrollment": false,
> > >   "household_size": "stuff",
> > >   "last_name": "Doe",
> > >   "phone_number": "224-705-2345",
> > >   "zip_code": "12345"
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The errors that are returned are about type, but does not include an error about missing the required field. Is this expected?
> > 
> > 
> > Hmm, this is due to how it's currently handled. We have validation in two places, the specification in openapi and the pydantic. Openapi happens first and if it encounters any errors, just returns those (going through some of our reformatting/logging). If it passes, then it hits the pydantic errors.
> > So, the type checking is happening in your request at the openapi layer, but we don't have anything marked as required at the moment on the request and pydantic handles that. This is the pattern Mass had, but we could add that to openapi so they all happen at once.
> 
> Ah I suspected that it was something like thate where we are doing two layes of checking and returning early if the first layer fails. This probably doesn't surprise you since I think I made a similar request with payments stuff, but I think it's a better API quality of life if all the errors are returned, rather than a subset.

Yeah, feels weird to have that split. Should we fix it here, or follow-up? I've always thought the pydantic validation was a little odd (beyond more complex business rules) ontop of the openapi validation.

**Review by rocketnova — APPROVED** [review]
LGTM!

**Comment by rocketnova** [comment]
> Yeah, feels weird to have that split. Should we fix it here, or follow-up? I've always thought the pydantic validation was a little odd (beyond more complex business rules) ontop of the openapi validation.

Let's do it in a follow-up. Can you add it to your list?
